backend: empty sql result sets not captured by sql.ErrNoRows checks for mysql dialect

### Checklist

- [X] I could not find a solution in the existing issues or docs.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/teamhanko/hanko/blob/main/CODE_OF_CONDUCT.md).

### Describe the bug

The backend has [several](https://github.com/teamhanko/hanko/search?q=sql.ErrNoRows) checks on whether a [pop](https://github.com/gobuffalo/buffalo-pop) error represents an empty result (`err == sql.ErrNoRows`). This check will return false when using mysql or mariadb/the mysql dialect. This leads to erroneous API responses on several endpoints, in fact it is possible that the application won't even start up due to the creation of the JWKs: 

When using the mysql dialect [this](https://github.com/teamhanko/hanko/blob/558caebb1f98a8b2e24e19978915afb99fb53888/backend/persistence/jwk_persister.go#L27) error will be an `fmt.WrapError` with a stringified value of `mysql select one: sql: no rows in result set` and hence does not [equal](https://github.com/teamhanko/hanko/blob/558caebb1f98a8b2e24e19978915afb99fb53888/backend/persistence/jwk_persister.go#L28) the value `sql: no rows in result set` of `sql.ErrNoRows`. This results in the JWK manager returning an error, then leading to a panic on startup instead of properly generating the JWKS.

Tested with (see also 'Reproducing the bug' below):

- MySQL 5.7
- MySQL 8.0.29
- MariaDB 10.9.3


### Reproducing the bug

1. Install docker
2. Run (one of):

```
# MySQL 5.7
docker run --name mysql57 -e MYSQL_USER=hanko -e MYSQL_PASSWORD=hanko -e MYSQL_DATABASE=hanko -e MYSQL_ROOT_PASSWORD=root -p 5432:3306 -d mysql:5.7

# MySQL 8.0.29 - latest tag version as of writing
docker run --name mysqllatest -e MYSQL_USER=hanko -e MYSQL_PASSWORD=hanko -e MYSQL_DATABASE=hanko -e MYSQL_ROOT_PASSWORD=root -p 5432:3306 -d mysql:latest

# MariaDB 10.9.3 - latest tag version as of writing
docker run --name mariadblatest -e MARIADB_USER=hanko -e MARIADB_PASSWORD=hanko -e MARIADB_DATABASE=hanko -e MARIADB_ROOT_PASSWORD=root -p 5432:3306 -d mariadb:latest
```
3. In `backend/config/config.yaml` change `database.dialect` to `mysql`
4. Apply migrations: `go run main.go migrate up` in `backend` dir
5. Run the backend: `go run main.go serve public` in `backend` dir

### Logs

```Shell
❯ go run main.go serve public
2022/10/26 17:34:48 Using config file: ./config/config.yaml
panic: failed to create jwk manager: failed to get jwk: mysql select one: sql: no rows in result set

goroutine 26 [running]:
github.com/teamhanko/hanko/backend/server.NewPublicRouter(0x204ca80, {0x999e098, 0xc00064c4f0})
        /Users/lfl/Documents/Hanko/hanko/backend/server/public_router.go:41 +0x18f9
github.com/teamhanko/hanko/backend/server.StartPublic(0x204ca80, 0x0, {0x999e098, 0xc00064c4f0})
        /Users/lfl/Documents/Hanko/hanko/backend/server/server.go:11 +0x6c
created by github.com/teamhanko/hanko/backend/cmd/serve.NewServePublicCommand.func1
        /Users/lfl/Documents/Hanko/hanko/backend/cmd/serve/public.go:29 +0x185
exit status 2
```


### Configuration

```yml
database:
  user: hanko
  password: hanko
  host: localhost
  port: 5432
  dialect: mysql
```


### Hanko Version

558caebb1f98a8b2e24e19978915afb99fb53888

### OS

macOS

### OS Version

macOS Monterey 12.6

### Browser Version

_No response_

### Environment

Binary/Build & Run from Source

### Additional Context

Possible solution: 

Instead of checking on `err == sql.ErrNoRows` we could/should probably use [`errors.Is`](https://pkg.go.dev/errors#Is) because it repeatedly unwraps errors (hence eventually leading to the `sql: no rows in result set` root cause).

--- a/backend/persistence/audit_log_persister.go
+++ b/backend/persistence/audit_log_persister.go
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"database/sql"
+	"errors"
 	"fmt"
 	"github.com/gobuffalo/pop/v6"
 	"github.com/gofrs/uuid"
@@ -39,7 +40,7 @@ func (p *auditLogPersister) Create(auditLog models.AuditLog) error {
 func (p *auditLogPersister) Get(id uuid.UUID) (*models.AuditLog, error) {
 	auditLog := models.AuditLog{}
 	err := p.db.Eager().Find(&auditLog, id)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {
@@ -61,7 +62,7 @@ func (p *auditLogPersister) List(page int, perPage int, startTime *time.Time, en
 	}
 	err := query.Paginate(page, perPage).Order("created_at desc").All(&auditLogs)
 
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return auditLogs, nil
 	}
 	if err != nil {

--- a/backend/persistence/jwk_persister.go
+++ b/backend/persistence/jwk_persister.go
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"database/sql"
+	"errors"
 	"fmt"
 	"github.com/gobuffalo/pop/v6"
 	"github.com/teamhanko/hanko/backend/persistence/models"
@@ -25,7 +26,7 @@ func NewJwkPersister(db *pop.Connection) JwkPersister {
 func (p *jwkPersister) Get(id int) (*models.Jwk, error) {
 	jwk := models.Jwk{}
 	err := p.db.Find(&jwk, id)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {
@@ -37,7 +38,7 @@ func (p *jwkPersister) Get(id int) (*models.Jwk, error) {
 func (p *jwkPersister) GetAll() ([]models.Jwk, error) {
 	jwks := []models.Jwk{}
 	err := p.db.All(&jwks)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {
@@ -49,7 +50,7 @@ func (p *jwkPersister) GetAll() ([]models.Jwk, error) {
 func (p *jwkPersister) GetLast() (*models.Jwk, error) {
 	jwk := models.Jwk{}
 	err := p.db.Last(&jwk)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {

--- a/backend/persistence/passcode_persister.go
+++ b/backend/persistence/passcode_persister.go
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"database/sql"
+	"errors"
 	"fmt"
 	"github.com/gobuffalo/pop/v6"
 	"github.com/gofrs/uuid"
@@ -26,7 +27,7 @@ func NewPasscodePersister(db *pop.Connection) PasscodePersister {
 func (p *passcodePersister) Get(id uuid.UUID) (*models.Passcode, error) {
 	passcode := models.Passcode{}
 	err := p.db.Find(&passcode, id)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {

--- a/backend/persistence/password_credential_persister.go
+++ b/backend/persistence/password_credential_persister.go
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"database/sql"
+	"errors"
 	"fmt"
 	"github.com/gobuffalo/pop/v6"
 	"github.com/gofrs/uuid"
@@ -39,7 +40,7 @@ func (p *passwordCredentialPersister) GetByUserID(userId uuid.UUID) (*models.Pas
 	pw := models.PasswordCredential{}
 	query := p.db.Where("user_id = (?)", userId.String())
 	err := query.First(&pw)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {

--- a/backend/persistence/user_persister.go
+++ b/backend/persistence/user_persister.go
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"database/sql"
+	"errors"
 	"fmt"
 	"github.com/gobuffalo/pop/v6"
 	"github.com/gofrs/uuid"
@@ -28,7 +29,7 @@ func NewUserPersister(db *pop.Connection) UserPersister {
 func (p *userPersister) Get(id uuid.UUID) (*models.User, error) {
 	user := models.User{}
 	err := p.db.Eager().Find(&user, id)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {
@@ -42,7 +43,7 @@ func (p *userPersister) GetByEmail(email string) (*models.User, error) {
 	user := models.User{}
 	query := p.db.Eager().Where("email = (?)", email)
 	err := query.First(&user)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {
@@ -91,6 +92,9 @@ func (p *userPersister) List(page int, perPage int) ([]models.User, error) {
 	users := []models.User{}
 
 	err := p.db.Q().Paginate(page, perPage).All(&users)
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return users, nil
+	}
 	if err != nil {
 		return nil, fmt.Errorf("failed to fetch users: %w", err)
 	}

--- a/backend/persistence/webauthn_credential_persister.go
+++ b/backend/persistence/webauthn_credential_persister.go
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"database/sql"
+	"errors"
 	"fmt"
 	"github.com/gobuffalo/pop/v6"
 	"github.com/gofrs/uuid"
@@ -27,7 +28,7 @@ func NewWebauthnCredentialPersister(db *pop.Connection) WebauthnCredentialPersis
 func (p *webauthnCredentialPersister) Get(id string) (*models.WebauthnCredential, error) {
 	credential := models.WebauthnCredential{}
 	err := p.db.Find(&credential, id)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {
@@ -88,7 +89,7 @@ func (p *webauthnCredentialPersister) Delete(credential models.WebauthnCredentia
 func (p *webauthnCredentialPersister) GetFromUser(userId uuid.UUID) ([]models.WebauthnCredential, error) {
 	var credentials []models.WebauthnCredential
 	err := p.db.Eager().Where("user_id = ?", &userId).All(&credentials)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {

--- a/backend/persistence/webauthn_session_data_persister.go
+++ b/backend/persistence/webauthn_session_data_persister.go
@@ -2,6 +2,7 @@ package persistence
 
 import (
 	"database/sql"
+	"errors"
 	"fmt"
 	"github.com/gobuffalo/pop/v6"
 	"github.com/gofrs/uuid"
@@ -27,7 +28,7 @@ func NewWebauthnSessionDataPersister(db *pop.Connection) WebauthnSessionDataPers
 func (p *webauthnSessionDataPersister) Get(id uuid.UUID) (*models.WebauthnSessionData, error) {
 	sessionData := models.WebauthnSessionData{}
 	err := p.db.Eager().Find(&sessionData, id)
-	if err != nil && err == sql.ErrNoRows {
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
 		return nil, nil
 	}
 	if err != nil {
@@ -40,6 +41,9 @@ func (p *webauthnSessionDataPersister) Get(id uuid.UUID) (*models.WebauthnSessio
 func (p *webauthnSessionDataPersister) GetByChallenge(challenge string) (*models.WebauthnSessionData, error) {
 	var sessionData []models.WebauthnSessionData
 	err := p.db.Eager().Where("challenge = ?", challenge).All(&sessionData)
+	if err != nil && errors.Is(err, sql.ErrNoRows) {
+		return nil, nil
+	}
 	if err != nil {
 		return nil, fmt.Errorf("failed to get sessionData: %w", err)
 	}

